File: src/Fable.Ripple/Internal/Graph.fs
namespace Fable.Ripple.Internal

open Fable.Ripple

/// Dependency edges with an inline-first-edge layout. A node keeps its first
/// source/observer in a scalar field and allocates the `Rest*` overflow
/// `ResizeArray` only on a second edge (holding logical indices 1..). These
/// helpers are the ONLY place that knows the layout - every other module goes
/// through them, so the "no `Rest*` without a `First*`" invariant lives here.
module internal Graph =

    let sourceCount (n: ReactiveNode) : int =
        match n.FirstSource with
        | ValueNone -> 0
        | ValueSome _ ->
            match n.RestSources with
            | ValueSome a -> 1 + a.Count
            | ValueNone -> 1

    /// Logical source `i` (caller guarantees `0 <= i < sourceCount n`).
    let sourceAt (n: ReactiveNode) (i: int) : ReactiveNode =
        if i = 0 then
            n.FirstSource.Value
        else
            n.RestSources.Value.[i - 1]

    let private ensureRestSources (n: ReactiveNode) =
        match n.RestSources with
        | ValueSome a -> a
        | ValueNone ->
            let a = ResizeArray<ReactiveNode>()
            n.RestSources <- ValueSome a
            a

    let addSource (n: ReactiveNode) (s: ReactiveNode) =
        match n.FirstSource with
        | ValueNone -> n.FirstSource <- ValueSome s
        | ValueSome _ -> (ensureRestSources n).Add s

    /// Keep the first `len` logical sources, drop the rest (keeps the overflow
    /// array allocated but empty, so a stable regrow reuses it).
    let truncateSources (n: ReactiveNode) (len: int) =
        if len <= 0 then
            n.FirstSource <- ValueNone
            n.RestSources |> ValueOption.iter (fun a -> a.Clear())
        else
            n.RestSources
            |> ValueOption.iter (fun a ->
                while a.Count > len - 1 do
                    a.RemoveAt(a.Count - 1)
            )

    /// Apply `action` to each logical source at index `>= fromIdx`.
    let inline iterSourcesFrom
        (n: ReactiveNode)
        (fromIdx: int)
        ([<InlineIfLambda>] action: ReactiveNode -> unit)
        =
        if fromIdx <= 0 then
            n.FirstSource |> ValueOption.iter action

        n.RestSources
        |> ValueOption.iter (fun a ->
            let start =
                if fromIdx <= 1 then
                    0
                else
                    fromIdx - 1

            for i in start .. a.Count - 1 do
                action a.[i]
        )

    /// Length of the observer list, including disposed entries not yet swept.
    let private observerSlots (n: ReactiveNode) : int =
        match n.FirstObserver with
        | ValueNone -> 0
        | ValueSome _ ->
            match n.RestObservers with
            | ValueSome a -> 1 + a.Count
            | ValueNone -> 1

    /// Live observers of `n` (disposed entries awaiting a sweep are not counted).
    let observerCount (n: ReactiveNode) : int = observerSlots n - n.DeadObservers

    let private ensureRestObservers (n: ReactiveNode) =
        match n.RestObservers with
        | ValueSome a -> a
        | ValueNone ->
            let a = ResizeArray<ReactiveNode>()
            n.RestObservers <- ValueSome a
            a

    let addObserver (n: ReactiveNode) (o: ReactiveNode) =
        match n.FirstObserver with
        | ValueNone -> n.FirstObserver <- ValueSome o
        | ValueSome _ -> (ensureRestObservers n).Add o

    /// Skips disposed observers; they stay listed until `sweepDeadObservers`.
    let inline iterObservers (n: ReactiveNode) ([<InlineIfLambda>] action: ReactiveNode -> unit) =
        n.FirstObserver
        |> ValueOption.iter (fun o ->
            if not o.Disposed then
                action o
        )

        n.RestObservers
        |> ValueOption.iter (fun a ->
            for i in 0 .. a.Count - 1 do
                let o = a.[i]

                if not o.Disposed then
                    action o
        )

    /// Swap-remove `o` from `n`'s observers (order is irrelevant).
    let removeObserver (n: ReactiveNode) (o: ReactiveNode) =
        match n.FirstObserver with
        | ValueSome f when obj.ReferenceEquals(f, o) ->
            match n.RestObservers with
            | ValueSome a when a.Count > 0 ->
                n.FirstObserver <- ValueSome a.[a.Count - 1]
                a.RemoveAt(a.Count - 1)
            | _ -> n.FirstObserver <- ValueNone
        | ValueSome _ ->
            n.RestObservers
            |> ValueOption.iter (fun a ->
                let mutable i = 0
                let mutable go = true

                while go && i < a.Count do
                    if obj.ReferenceEquals(a.[i], o) then
                        a.[i] <- a.[a.Count - 1]
                        a.RemoveAt(a.Count - 1)
                        go <- false
                    else
                        i <- i + 1
            )
        | ValueNone -> ()

    /// Keep only the observers of `n` satisfying `keep`, compacting in place with
    /// no allocation: the first kept edge fills `FirstObserver` (promoting one out
    /// of the overflow if needed), the rest are compacted into `RestObservers` by a
    /// write cursor. `w <= readIdx` always, so writes never clobber unread slots.
    let compactObservers (n: ReactiveNode) (keep: ReactiveNode -> bool) =
        match n.FirstObserver with
        | ValueNone -> ()
        | ValueSome f0 ->
            let mutable newFirst =
                if keep f0 then
                    ValueSome f0
                else
                    ValueNone

            n.RestObservers
            |> ValueOption.iter (fun rest ->
                let mutable w = 0

                for readIdx in 0 .. rest.Count - 1 do
                    let o = rest.[readIdx]

                    if keep o then
                        match newFirst with
                        | ValueNone -> newFirst <- ValueSome o
                        | ValueSome _ ->
                            rest.[w] <- o
                            w <- w + 1

                while rest.Count > w do
                    rest.RemoveAt(rest.Count - 1)
            )

            n.FirstObserver <- newFirst

    /// Threshold of half: a sweep removes at least as many entries as it keeps,
    /// so sweeping stays linear in disposals.
    let private sweepDeadObservers (n: ReactiveNode) =
        let dead = n.DeadObservers

        if dead > 0 then
            let slots = observerSlots n

            if dead >= slots then
                n.FirstObserver <- ValueNone
                n.RestObservers <- ValueNone
                n.DeadObservers <- 0
            elif 2 * dead >= slots then
                compactObservers n (fun o -> not o.Disposed)
                n.DeadObservers <- 0

    // `Affected` marks membership in `sweepQueue`.
    let private sweepQueue = ResizeArray<ReactiveNode>()
    let mutable private sweepHolds = 0

    let noteDeadObserver (source: ReactiveNode) =
        source.DeadObservers <- source.DeadObservers + 1

        if not source.Affected then
            source.Affected <- true
            sweepQueue.Add source

    /// Nests; the outermost `releaseSweeps` runs the queued checks.
    let holdSweeps () = sweepHolds <- sweepHolds + 1

    /// Release a hold; the last one runs the queued sweep checks.
    let releaseSweeps () =
        sweepHolds <- sweepHolds - 1

        if sweepHolds = 0 && sweepQueue.Count > 0 then
            for i in 0 .. sweepQueue.Count - 1 do
                let source = sweepQueue.[i]
                source.Affected <- false
                sweepDeadObservers source

            sweepQueue.Clear()

    /// Unlink `node` from each of its sources at or after `fromIndex`.
    let unlinkSourcesTail (node: ReactiveNode) (fromIndex: int) =
        iterSourcesFrom node fromIndex (fun s -> removeObserver s node)

    /// Unlink a node from the graph entirely (used to dispose subscriptions).
    let dispose (node: ReactiveNode) =
        iterSourcesFrom node 0 (fun s -> removeObserver s node)
        node.FirstSource <- ValueNone
        node.RestSources |> ValueOption.iter (fun a -> a.Clear())
        node.State <- NodeState.Clean
        node.Queued <- false

File: visuals/src/Fable.Ripple/Internal/Graph.js

import { ReactiveNode__set_Queued_Z1FBCCD16, ReactiveNode__set_State_Z12CE0414, ReactiveNode__set_Affected_Z1FBCCD16, ReactiveNode__get_Affected, ReactiveNode__get_Disposed, ReactiveNode__set_DeadObservers_Z524259A4, ReactiveNode__set_FirstObserver_Z46457FEC, ReactiveNode__set_RestObservers_6EF2C44D, ReactiveNode__get_DeadObservers, ReactiveNode__get_RestObservers, ReactiveNode__get_FirstObserver, ReactiveNode__set_FirstSource_Z46457FEC, ReactiveNode__set_RestSources_6EF2C44D, ReactiveNode__get_RestSources, ReactiveNode__get_FirstSource } from "../ReactiveNode.js";
import { value } from "../../../fable_modules/fable-library-js.5.13.0/Option.js";
import { setItem, item } from "../../../fable_modules/fable-library-js.5.13.0/Array.js";
import { clear } from "../../../fable_modules/fable-library-js.5.13.0/Util.js";

export function sourceCount(n) {
    const matchValue = ReactiveNode__get_FirstSource(n);
    if (matchValue != null) {
        const matchValue_1 = ReactiveNode__get_RestSources(n);
        if (matchValue_1 == null) {
            return 1;
        }
        else {
            const a = matchValue_1;
            return (1 + a.length) | 0;
        }
    }
    else {
        return 0;
    }
}

/**
 * Logical source `i` (caller guarantees `0 <= i < sourceCount n`).
 */
export function sourceAt(n, i) {
    let copyOfStruct_1;
    if (i === 0) {
        let copyOfStruct = ReactiveNode__get_FirstSource(n);
        return value(copyOfStruct);
    }
    else {
        return item(i - 1, (copyOfStruct_1 = ReactiveNode__get_RestSources(n), value(copyOfStruct_1)));
    }
}

function ensureRestSources(n) {
    const matchValue = ReactiveNode__get_RestSources(n);
    if (matchValue == null) {
        const a_1 = [];
        ReactiveNode__set_RestSources_6EF2C44D(n, a_1);
        return a_1;
    }
    else {
        return matchValue;
    }
}

export function addSource(n, s) {
    const matchValue = ReactiveNode__get_FirstSource(n);
    if (matchValue != null) {
        void (ensureRestSources(n).push(s));
    }
    else {
        ReactiveNode__set_FirstSource_Z46457FEC(n, s);
    }
}

/**
 * Keep the first `len` logical sources, drop the rest (keeps the overflow
 * array allocated but empty, so a stable regrow reuses it).
 */
export function truncateSources(n, len) {
    if (len <= 0) {
        ReactiveNode__set_FirstSource_Z46457FEC(n, undefined);
        const option = ReactiveNode__get_RestSources(n);
        if (option != null) {
            clear(option);
        }
    }
    else {
        const option_1 = ReactiveNode__get_RestSources(n);
        if (option_1 != null) {
            const a_1 = option_1;
            while (a_1.length > (len - 1)) {
                a_1.splice(a_1.length - 1, 1);
            }
        }
    }
}

function observerSlots(n) {
    const matchValue = ReactiveNode__get_FirstObserver(n);
    if (matchValue != null) {
        const matchValue_1 = ReactiveNode__get_RestObservers(n);
        if (matchValue_1 == null) {
            return 1;
        }
        else {
            const a = matchValue_1;
            return (1 + a.length) | 0;
        }
    }
    else {
        return 0;
    }
}

/**
 * Live observers of `n` (disposed entries awaiting a sweep are not counted).
 */
export function observerCount(n) {
    return (observerSlots(n) - ReactiveNode__get_DeadObservers(n)) | 0;
}

function ensureRestObservers(n) {
    const matchValue = ReactiveNode__get_RestObservers(n);
    if (matchValue == null) {
        const a_1 = [];
        ReactiveNode__set_RestObservers_6EF2C44D(n, a_1);
        return a_1;
    }
    else {
        return matchValue;
    }
}

export function addObserver(n, o) {
    const matchValue = ReactiveNode__get_FirstObserver(n);
    if (matchValue != null) {
        void (ensureRestObservers(n).push(o));
    }
    else {
        ReactiveNode__set_FirstObserver_Z46457FEC(n, o);
    }
}

/**
 * Swap-remove `o` from `n`'s observers (order is irrelevant).
 */
export function removeObserver(n, o) {
    let a;
    const matchValue = ReactiveNode__get_FirstObserver(n);
    if (matchValue == null) {
    }
    else if (matchValue === o) {
        const matchValue_1 = ReactiveNode__get_RestObservers(n);
        let matchResult;
        if (matchValue_1 != null) {
            if ((a = matchValue_1, a.length > 0)) {
                matchResult = 0;
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
        switch (matchResult) {
            case 0: {
                const a_1 = matchValue_1;
                ReactiveNode__set_FirstObserver_Z46457FEC(n, item(a_1.length - 1, a_1));
                a_1.splice(a_1.length - 1, 1);
                break;
            }
            case 1: {
                ReactiveNode__set_FirstObserver_Z46457FEC(n, undefined);
                break;
            }
        }
    }
    else {
        const option = ReactiveNode__get_RestObservers(n);
        if (option != null) {
            const a_2 = option;
            let i = 0;
            let go = true;
            while (go && (i < a_2.length)) {
                if (item(i, a_2) === o) {
                    setItem(a_2, i, item(a_2.length - 1, a_2));
                    a_2.splice(a_2.length - 1, 1);
                    go = false;
                }
                else {
                    i = ((i + 1) | 0);
                }
            }
        }
    }
}

/**
 * Keep only the observers of `n` satisfying `keep`, compacting in place with
 * no allocation: the first kept edge fills `FirstObserver` (promoting one out
 * of the overflow if needed), the rest are compacted into `RestObservers` by a
 * write cursor. `w <= readIdx` always, so writes never clobber unread slots.
 */
export function compactObservers(n, keep) {
    const matchValue = ReactiveNode__get_FirstObserver(n);
    if (matchValue != null) {
        const f0 = matchValue;
        let newFirst = keep(f0) ? f0 : undefined;
        const option = ReactiveNode__get_RestObservers(n);
        if (option != null) {
            const rest = option;
            let w = 0;
            for (let readIdx = 0; readIdx <= (rest.length - 1); readIdx++) {
                const o = item(readIdx, rest);
                if (keep(o)) {
                    if (newFirst != null) {
                        setItem(rest, w, o);
                        w = ((w + 1) | 0);
                    }
                    else {
                        newFirst = o;
                    }
                }
            }
            while (rest.length > w) {
                rest.splice(rest.length - 1, 1);
            }
        }
        ReactiveNode__set_FirstObserver_Z46457FEC(n, newFirst);
    }
}

function sweepDeadObservers(n) {
    const dead = ReactiveNode__get_DeadObservers(n) | 0;
    if (dead > 0) {
        const slots = observerSlots(n) | 0;
        if (dead >= slots) {
            ReactiveNode__set_FirstObserver_Z46457FEC(n, undefined);
            ReactiveNode__set_RestObservers_6EF2C44D(n, undefined);
            ReactiveNode__set_DeadObservers_Z524259A4(n, 0);
        }
        else if ((2 * dead) >= slots) {
            compactObservers(n, (o) => !ReactiveNode__get_Disposed(o));
            ReactiveNode__set_DeadObservers_Z524259A4(n, 0);
        }
    }
}

const sweepQueue = [];

let sweepHolds = 0;

export function noteDeadObserver(source) {
    ReactiveNode__set_DeadObservers_Z524259A4(source, ReactiveNode__get_DeadObservers(source) + 1);
    if (!ReactiveNode__get_Affected(source)) {
        ReactiveNode__set_Affected_Z1FBCCD16(source, true);
        void (sweepQueue.push(source));
    }
}

/**
 * Nests; the outermost `releaseSweeps` runs the queued checks.
 */
export function holdSweeps() {
    sweepHolds = ((sweepHolds + 1) | 0);
}

/**
 * Release a hold; the last one runs the queued sweep checks.
 */
export function releaseSweeps() {
    sweepHolds = ((sweepHolds - 1) | 0);
    if ((sweepHolds === 0) && (sweepQueue.length > 0)) {
        for (let i = 0; i <= (sweepQueue.length - 1); i++) {
            const source = item(i, sweepQueue);
            ReactiveNode__set_Affected_Z1FBCCD16(source, false);
            sweepDeadObservers(source);
        }
        clear(sweepQueue);
    }
}

/**
 * Unlink `node` from each of its sources at or after `fromIndex`.
 */
export function unlinkSourcesTail(node, fromIndex) {
    const n = node;
    const fromIdx = fromIndex | 0;
    if (fromIdx <= 0) {
        const option = ReactiveNode__get_FirstSource(n);
        if (option != null) {
            removeObserver(option, node);
        }
    }
    const option_1 = ReactiveNode__get_RestSources(n);
    if (option_1 != null) {
        const a = option_1;
        const start = ((fromIdx <= 1) ? 0 : (fromIdx - 1)) | 0;
        for (let i = start; i <= (a.length - 1); i++) {
            removeObserver(item(i, a), node);
        }
    }
}

/**
 * Unlink a node from the graph entirely (used to dispose subscriptions).
 */
export function dispose(node) {
    const n = node;
    if (0 <= 0) {
        const option = ReactiveNode__get_FirstSource(n);
        if (option != null) {
            removeObserver(option, node);
        }
    }
    const option_1 = ReactiveNode__get_RestSources(n);
    if (option_1 != null) {
        const a = option_1;
        const start = ((0 <= 1) ? 0 : (0 - 1)) | 0;
        for (let i = start; i <= (a.length - 1); i++) {
            removeObserver(item(i, a), node);
        }
    }
    ReactiveNode__set_FirstSource_Z46457FEC(node, undefined);
    const option_2 = ReactiveNode__get_RestSources(node);
    if (option_2 != null) {
        clear(option_2);
    }
    ReactiveNode__set_State_Z12CE0414(node, 0);
    ReactiveNode__set_Queued_Z1FBCCD16(node, false);
}

